client/consensus/grandpa/authorities.go

```diff
@@ -705,64 +704,33 @@ func (asc *AuthoritySetChanges[N]) append(setID uint64, blockNumber N) {
 	})
 }
 
-type authoritySetChangeID scale.VaryingDataType
-
-// Set will set a VaryingDataTypeValue using the underlying VaryingDataType
-func (asc *authoritySetChangeID) Set(val scale.VaryingDataTypeValue) (err error) {
-	vdt := scale.VaryingDataType(*asc)
-	err = vdt.Set(val)
-	if err != nil {
-		return
-	}
-	*asc = authoritySetChangeID(vdt)
-	return
-}
-
-// Value will return value from underying VaryingDataType
-func (asc *authoritySetChangeID) Value() (val scale.VaryingDataTypeValue, err error) {
-	vdt := scale.VaryingDataType(*asc)
-	return vdt.Value()
+type authoritySetChangeID interface {
```

Comment:
It was per suggestion from this comment, https://github.com/ChainSafe/gossamer/pull/3589#discussion_r1408247589 and no I do not believe that it is encoded. Will try your suggestion

Reply:
updated

Reply:
Changing `authoritySetChangeID` from an interface to any and introducing a constructor that returns `any` is not worth. It adds complexity but in the end, is the same as simply returning any. I don't get the point of a constructor that instead of construct type is returning any? So I would keep suggesting https://github.com/ChainSafe/gossamer/pull/3589#discussion_r1408247589, also if we want to implement enums that don't need to be scale encoded decoded we should use interfaces, take as example https://github.com/ChainSafe/gossamer/pull/3570

Reply:
`any` is the same as `interface{}`.  I don't see the need to create a superfluous interface to represent these 3 options when they can be represented by an interface constraint.  By creating a  constructor that uses this constraint as a generic parameter we can enforce what is stored as a `authoritySetChangeID` at runtime as well.

Reply:
@timwu20 @jimjbrettj  I wrote a simple post on hackmd with a cleaner example and explanation about why using the suggestion I made is not superfluous and has a better type-safety and less complex code and why have a constructor don't have those guarantees

https://hackmd.io/@-8f0KkDaSXGl-AgsJkF8Jg/B1X1WAxIT

Reply:
This approach makes sense to me, thank you for the write-up @EclesioMeloJunior

Reply:
I have updated the code to reflect your suggestion @EclesioMeloJunior

Reply:
This approach is not necessary.  You can ensure type safety through any function by using the interface type constraint.  See this [play link](https://go.dev/play/p/6g58lCu2yir).

Reply:
Not in this case since to use interface type constraint we will need to use `any` as a "hack" to support the type constraint enum as a "shadowed" value, I will keep suggesting the approach described here https://github.com/ChainSafe/gossamer/pull/3589#discussion_r1408247589. ref: https://zephyrtronium.github.io/articles/generics-enums.html

Reply:
Based on our conversations on slack, I think there's an acceptable tradeoff here to use an interface constraint here to constrain what we store in the shadowed `any` type.  Using an interface constraint makes it clear the intention here.  Hopefully they introduce something in future versions of Go to be able to replicate sum types (aka. rust enums).

Reply:
This sounds good to me and I will keep this suggestion in my code and mark this as resolved.